wa/workloads/recentfling/__init__.py

```diff
@@ -0,0 +1,126 @@
+    in ``~/.workload_automation/dependencies/recentfling/``. These can be found
+    in the `AOSP Git repository <https://android.googlesource.com/platform/system/extras/+/master/tests/workloads>`_.
+
+    To change the apps that are opened at the start of the workload you will need
+    to modify the ``defs.sh`` file. You will need to add your app to ``dfltAppList``
+    and then add a variable called ``{app_name}Activity`` with the name of the
+    activity to launch (where ``{add_name}`` is the name you put into ``dfltAppList``).
+
+    You can get a list of activities available on your device by running
+    ``adb shell pm list packages -f``
+    """
+    supported_platforms = ['android']
+
+    parameters = [
+        Parameter('loops', kind=int, default=3,
+                  description="The number of test iterations."),
+        Parameter('start_apps', kind=bool, default=True,
+                  description="""
+                  If set to ``False``,no apps will be started before flinging
+                  through the recent apps list (in which the assumption is
+                  there are already recently started apps in the list.
+                  """),
+        Parameter('device_name', kind=str,
+                  description="""
+                  If set, recentfling will use the fling parameters for this
+                  device instead of automatically guessing the device.  This can
+                  also be used if the device is not supported by recentfling,
+                  but its screensize is similar to that of one that is supported.
+
+                  For possible values, check your recentfling.sh.  At the time
+                  of writing, valid values are: 'shamu', 'hammerhead', 'angler',
+                  'ariel', 'mtp8996', 'bullhead' or 'volantis'.
+                  """),
+    ]
+
+    def initialise(self, context):  # pylint: disable=no-self-use
+        if context.target.get_sdk_version() < 23:
+            raise WorkloadError("This workload relies on ``dumpsys gfxinfo`` \
+                                 only present in Android M and onwards")
+
+    def setup(self, context):
+        self.defs_host = context.resolver.get(File(self, "defs.sh"))
+        self.defs_target = self.target.install(self.defs_host)
+        self.recentfling_host = context.resolver.get(File(self, "recentfling.sh"))
+        self.recentfling_target = self.target.install(self.recentfling_host)
+        self._kill_recentfling()
+        self.target.ensure_screen_is_on()
+
+    def run(self, context):
+        args = '-i {} '.format(self.loops)
+        if not self.start_apps:
+            args += '-N '
+        if self.device_name is not None:
+            args += '-d {}'.format(self.device_name)
+        cmd = "echo $$>{dir}/pidfile; cd {bindir}; exec ./recentfling.sh {args}; rm {dir}/pidfile"
+        cmd = cmd.format(args=args, dir=self.target.working_directory,
+                         bindir=self.target.executables_directory)
```

Comment:
Everything up to this point should be moved into the `setup`; `run` should be kept as small  as possible.

Reply:
Why `setup()` and not `initialize()`?  If we are going to move it, we may as well put it in `initialize()`.